cql3/statements/strongly_consistent_modification_statement.cc

```diff
@@ -0,0 +1,147 @@
+    }
+
+    return to_bytes(value->view());
+}
+
+static
+bytes get_new_value(const std::vector<::shared_ptr<operation>>& operations) {
+    if (operations.size() != 1) {
+        throw raft::broadcast_tables::unsupported_operation_error("modifications: {}", fmt::join(
+            operations | transformed([] (const auto& op) { return fmt::format("{}", op->get_expression()); }), ", "));
+    }
+
+    const auto optional_expression = operations[0]->get_expression();
+
+    if (!optional_expression) {
+        throw raft::broadcast_tables::unsupported_operation_error("modifications: {}", fmt::join(
+            operations | transformed([] (const auto& op) { return fmt::format("{}", op->get_expression()); }), ", "));
+    }
+
+    const auto* value = cql3::expr::as_if<cql3::expr::constant>(&*optional_expression);
+
+    if (!value) {
+        throw raft::broadcast_tables::unsupported_operation_error("modification: {}", optional_expression);
+    }
+
+    return to_bytes(value->view());
+}
+
+static
+std::optional<bytes_opt> get_value_condition(const std::vector<lw_shared_ptr<column_condition>>& conditions) {
+    if (conditions.size() == 0) {
+        return std::nullopt;
+    }
+
+    if (conditions.size() > 1) {
+        throw raft::broadcast_tables::unsupported_operation_error("conditions: {}", fmt::join(
+            conditions | transformed([] (const auto& cond) {
+                return fmt::format("{}{}", cond->get_operation(), cond->get_value());
+            }), ", "));
+    }
+
+    const auto& condition = conditions[0];
+
+    if (!condition->get_value() || condition->get_operation() != cql3::expr::oper_t::EQ) {
+        throw raft::broadcast_tables::unsupported_operation_error("condition: {}{}", condition->get_operation(), condition->get_value());
+    }
+
+    const auto* value = cql3::expr::as_if<cql3::expr::constant>(&*condition->get_value());
+
+    if (!value) {
+        throw raft::broadcast_tables::unsupported_operation_error("condition: {}{}", condition->get_operation(), condition->get_value());
+    }
+
+    if (value->is_null()) {
+        return bytes_opt{};
+    }
+
+    return to_bytes(value->view());
+}
+
+void strongly_consistent_modification_statement::prepare_raft_command() {
+    _query.emplace(raft::broadcast_tables::update_query{
+        .key = get_key(restrictions().get_partition_key_restrictions()),
+        .new_value = get_new_value(_column_operations),
+        .value_condition = get_value_condition(_regular_conditions),
+    });
+}
+
+bool strongly_consistent_modification_statement::require_full_clustering_key() const {
+    return true;
+}
+
+bool strongly_consistent_modification_statement::allow_clustering_key_slices() const {
+    return false;
+}
+
+void strongly_consistent_modification_statement::add_update_for_key(mutation& m, const query::clustering_range& range, const update_parameters& params, const json_cache_opt& json_cache) const {
```

Comment:
> What made you decide to make strongly_consistent_modification_statement a subclass of modification_statement in the first place, and not e.g. a subclass of cql_statement_opt_metadata (which is a much more generic interface, basically validate and execute)?

That's because the return type of `raw::update_statement::prepare_internal` is a pointer to `modification_statement`, thus if we want `strongly_consistent_modification_statement` to be a subclass of `cql_statement_opt_metadata`, the return type has to be changed (what I think is impossible) or we have to add a new subclass of `raw::cf_statement`.

Reply:
I really dislike the design of `modification_statement`.
First you see this:
```
/*
 * Abstract parent class of individual modifications, i.e. INSERT, UPDATE and DELETE.
 */
class modification_statement : public cql_statement_opt_metadata {
```

you'd expect a simple interface with one or two virtual methods.

But then you read further and see all these concrete (and public) member functions, which are basically implementation details of the subclasses. So now when we want to add another modification statement which executes updates a bit differently, we're left with all this stuff that we don't need and we have to implement empty overrides of virtual members. This whole class is one big antipattern, doing "composition" by inheritance.

Anyway, enough with the ranting.

> the return type has to be changed (what I think is impossible)

Can you explain why you think it's impossible?

Does the caller of `prepare_internal` use the concrete type information? If they do - what is it used for? Maybe this could be refactored. The purpose of `prepare` should be to return an abstract statement type which offers a way to bind variables and to execute it. Indeed, `prepare` returns a `prepared_statement` which contains `shared_ptr<cql_statement>` inside and `cql_statement` is very abstract. I'm wondering why we would need to return anything more concrete at any point.

Reply:
> I really dislike the design of `modification_statement`. First you see this:
> 
> ```
> /*
>  * Abstract parent class of individual modifications, i.e. INSERT, UPDATE and DELETE.
>  */
> class modification_statement : public cql_statement_opt_metadata {
> ```
> 
> you'd expect a simple interface with one or two virtual methods.
> 
> But then you read further and see all these concrete (and public) member functions, which are basically implementation details of the subclasses. So now when we want to add another modification statement which executes updates a bit differently, we're left with all this stuff that we don't need and we have to implement empty overrides of virtual members. This whole class is one big antipattern, doing "composition" by inheritance.

Agree with this criticism. Sharing implementation details should be done via helper functions, or via a further subclass.

> 
> Anyway, enough with the ranting.
> 
> > the return type has to be changed (what I think is impossible)
> 
> Can you explain why you think it's impossible?
> 
> Does the caller of `prepare_internal` use the concrete type information? If they do - what is it used for? Maybe this could be refactored. The purpose of `prepare` should be to return an abstract statement type which offers a way to bind variables and to execute it. Indeed, `prepare` returns a `prepared_statement` which contains `shared_ptr<cql_statement>` inside and `cql_statement` is very abstract. I'm wondering why we would need to return anything more concrete at any point.

It might be premature use of a covariant return type and likely can be degraded.

Reply:
> Does the caller of prepare_internal use the concrete type information? If they do - what is it used for?

`prepare_internal` is called inside `modification_statement::prepare`.  ` modification_statement::prepare` clears the `ctx` cache (for non-conditional statements) after calling `prepare_internal`.

Reply:
Maybe that can be moved to individual statement implementations.

Reply:
This piece of code looks like a good candidate to be moved into this `prepare_conditions` helper, which is called by `prepare_internal` implementations where applicable.

But be careful: some subclasses call `process_where_clause` after `prepare_conditions`, and it appears that this piece of code requires both `process_where_clause` and `prepare_conditions` to finish first (it accesses `_restrictions`, which is calculated in `process_where_clause` AFAIK).

Perhaps we could move `prepare_conditions` after `process_where_clause` in those implementations? And state in the interface of `prepare_conditions` that `_restrictions` must be calculated by this point, or something like that. Would have to check if there is no dependence in the other direction already to see if that's possible.

If we were able to move `prepare_conditions` after `process_where_clause,` then we could move this piece of code at the end of `prepare_conditions`.

Reply:
Next question is whether all of this is worth it.

Even if we manage to change the return type of `prepare_internal` to return a more abstract base class, there's still this part of `prepare_internal` happening after `create_statement()` call (`create_statement which you added), which depends on the concrete type returned by `create_statement`, which right now returns `modification_statement`. There are calls to `add_operation` and `prepare_conditions` and `process_where_clause`. I'm guessing our `strongly_consistent_modification_statement` needs some of this stuff too.

I see two approaches we can take:
Approach 1:
- refactor the part inside `prepare_internal` after `create_statement` into a virtual function, with one implementation for `strongly_consistent_modification_statement` and another implementation shared by the other statements
- the parts which both types of statements need would have to be refactored into helper objects/functions so there is no logic duplication
- note: this would also solve the `prepare_raft_comand` problem

Approach 2:
- ignore these design problems and live with the pain of subclassing `modification_statement`  and having empty implementations; `modification_statement` has a refactor coming for it eventually anyway :>

I would prefer approach 1, but you should investigate how much pain it would take to do the refactors, and how much refactoring is actually needed, and then decide for yourself (and convince us that it's too much pain).

Reply:
If you do decide to refactor, it's a good idea to retire it early in a separate pull request.

Reply:
`raw::modification_statement::prepare` is also called by `batch_statement` which stores a vector of `modification_statement`, thus we cannot simply change the return type of `prepare_internal`.

Reply:
I would choose the second approach. `prepare_conditions` can't be simply moved after `process_where_clause`. `process_where_clause` validates the primary key based on prepared conditions. It can be moved from `process_where_clause`, but then we will probably end up with a new virtual function with specialization only for one subclass (`insert_json_statement` doesn't call `process_where_clause` and doesn't have the primary key validation).

Reply:
`strongly_consistent_modification_statement` is now prepared by `modification_statement` and derives from `cql_statement_opt_metadata`. Preparation of `modification_statement` is reused: for every request, it is prepared and if it's broadcast tables query, it will prepare (`prepare_for_raft`) and return the raft command. `prepare_for_raft` is implemented only for `update_statement`, but it will be eventually (if we decide to choose active replication) implemented in others statements (for now, there is a default implementation that throws an exception).